pkg/policy/selectorcache_selector.go

```diff
@@ -0,0 +1,358 @@
+			return nil, err
+		}
+
+		buffer.Write(buf)
+		if i < len(s)-1 {
+			buffer.WriteString(",")
+		}
+	}
+	buffer.WriteString("]")
+	return buffer.Bytes(), nil
+}
+
+func (s CachedSelectorSlice) Len() int      { return len(s) }
+func (s CachedSelectorSlice) Swap(i, j int) { s[i], s[j] = s[j], s[i] }
+
+func (s CachedSelectorSlice) Less(i, j int) bool {
+	return strings.Compare(s[i].String(), s[j].String()) < 0
+}
+
+// SelectsAllEndpoints returns whether the CachedSelectorSlice selects all
+// endpoints, which is true if the wildcard endpoint selector is present in the
+// slice.
+func (s CachedSelectorSlice) SelectsAllEndpoints() bool {
+	for _, selector := range s {
+		if selector.IsWildcard() {
+			return true
+		}
+	}
+	return false
+}
+
+// CachedSelectionUser inserts selectors into the cache and gets update
+// callbacks whenever the set of selected numeric identities change for
+// the CachedSelectors pushed by it.
+type CachedSelectionUser interface {
+	// IdentitySelectionUpdated implementations MUST NOT call back
+	// to the name manager or the selector cache while executing this function!
+	//
+	// The caller is responsible for making sure the same identity is not
+	// present in both 'added' and 'deleted'.
+	IdentitySelectionUpdated(selector CachedSelector, added, deleted []identity.NumericIdentity)
+}
+
+// identitySelector is the internal type for all selectors in the
+// selector cache.
+//
+// identitySelector represents the mapping of an EndpointSelector
+// to a slice of identities. These mappings are updated via two
+// different processes:
+//
+// 1. When policy rules are changed these are added and/or deleted
+// depending on what selectors the rules contain. Cached selections of
+// new identitySelectors are pre-populated from the set of currently
+// known identities.
+//
+// 2. When reachacble identities appear or disappear, either via local
+// allocation (CIDRs), or via the KV-store (remote endpoints). In this
+// case all existing identitySelectors are walked through and their
+// cached selections are updated as necessary.
+//
+// In both of the above cases the set of existing identitySelectors is
+// write locked.
+//
+// To minimize the upkeep the identity selectors are shared across
+// all IdentityPolicies, so that only one copy exists for each
+// identitySelector. Users of the SelectorCache take care of creating
+// identitySelectors as needed by identity policies. The set of
+// identitySelectors is read locked during an IdentityPolicy update so
+// that the the policy is always updated using a coherent set of
+// cached selections.
+//
+// identitySelector is used as a map key, so it must not be implemented by a
+// map, slice, or a func, or a runtime panic will be triggered. In all
+// cases below identitySelector is being implemented by structs.
+//
+// Because identitySelector is exposed to the user is used as a map key, it must always
+// be passed to the user as a pointer to the actual type.
```

Comment:
I think this no longer applies now that `identitySelector` is no longer an interface, ref. d84ee5215c7a695a8c23bcfa41098d8860de58e5. Though it's probably still beneficial to pass as a pointer for different reasons.

Reply:
It does still apply, since the only exposed type is the CachedSelection interface. I'll clarify this in the doc.